File: benchmarks/StringThing.Benchmarks/InListBenchmarks.cs
using BenchmarkDotNet.Attributes;
using Dapper;
using Microsoft.Data.Sqlite;
using StringThing.Sqlite;
using StringThing.Sqlite.Dapper;

namespace StringThing.Benchmarks;

[MemoryDiagnoser]
public class InListBenchmarks
{
    private SqliteConnection _connection = null!;

    private static readonly string _prebuiltSql10 =
        "SELECT id, name, email FROM users WHERE id IN ("
        + string.Join(", ", Enumerable.Range(0, 10).Select(i => $"@p{i}"))
        + ")";

    private static readonly string _prebuiltSql100 =
        "SELECT id, name, email FROM users WHERE id IN ("
        + string.Join(", ", Enumerable.Range(0, 100).Select(i => $"@p{i}"))
        + ")";

    private int[] _tenItems = null!;
    private int[] _hundredItems = null!;
    private List<int> _tenItemsList = null!;
    private List<int> _hundredItemsList = null!;

    public record User(long Id, string Name, string? Email);

    [GlobalSetup]
    public void Setup()
    {
        _connection = new SqliteConnection("Data Source=:memory:");
        _connection.Open();

        using var create = _connection.CreateCommand();
        create.CommandText = """
            CREATE TABLE users (
                id INTEGER PRIMARY KEY,
                name TEXT NOT NULL,
                email TEXT
            )
            """;
        create.ExecuteNonQuery();

        for (var i = 1; i <= 100; i++)
        {
            using var insert = _connection.CreateCommand();
            insert.CommandText = $"INSERT INTO users (id, name, email) VALUES ({i}, 'user_{i}', 'user_{i}@example.com')";
            insert.ExecuteNonQuery();
        }

        _tenItems = Enumerable.Range(1, 10).ToArray();
        _hundredItems = Enumerable.Range(1, 100).ToArray();
        _tenItemsList = _tenItems.ToList();
        _hundredItemsList = _hundredItems.ToList();
    }

    [GlobalCleanup]
    public void Cleanup()
    {
        _connection.Dispose();
    }

    // --- 10 items ---

    [Benchmark(Description = "Raw: IN 10 (prebuilt SQL)")]
    public List<User> Raw_InList_Ten()
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = _prebuiltSql10;
        for (var i = 0; i < 10; i++)
            cmd.Parameters.AddWithValue($"@p{i}", _tenItems[i]);
        using var reader = cmd.ExecuteReader();
        var results = new List<User>();
        while (reader.Read())
            results.Add(new User(reader.GetInt64(0), reader.GetString(1), reader.IsDBNull(2) ? null : reader.GetString(2)));
        return results;
    }

    [Benchmark(Description = "Dapper: IN 10")]
    public List<User> Dapper_InList_Ten()
    {
        return _connection.Query<User>(
            "SELECT id, name, email FROM users WHERE id IN @ids",
            new { ids = _tenItemsList }).ToList();
    }

    [Benchmark(Description = "StringThing: IN 10")]
    public List<User> StringThing_InList_Ten()
    {
        return _connection.QueryString<User>(
            $"SELECT id, name, email FROM users WHERE id IN {SqliteSql.InList([.. _tenItems])}").ToList();
    }

    // --- 100 items ---

    [Benchmark(Description = "Raw: IN 100 (prebuilt SQL)")]
    public List<User> Raw_InList_Hundred()
    {
        using var cmd = _connection.CreateCommand();
        cmd.CommandText = _prebuiltSql100;
        for (var i = 0; i < 100; i++)
            cmd.Parameters.AddWithValue($"@p{i}", _hundredItems[i]);
        using var reader = cmd.ExecuteReader();
        var results = new List<User>();
        while (reader.Read())
            results.Add(new User(reader.GetInt64(0), reader.GetString(1), reader.IsDBNull(2) ? null : reader.GetString(2)));
        return results;
    }

    [Benchmark(Description = "Dapper: IN 100")]
    public List<User> Dapper_InList_Hundred()
    {
        return _connection.Query<User>(
            "SELECT id, name, email FROM users WHERE id IN @ids",
            new { ids = _hundredItemsList }).ToList();
    }

    [Benchmark(Description = "StringThing: IN 100")]
    public List<User> StringThing_InList_Hundred()
    {
        return _connection.QueryString<User>(
            $"SELECT id, name, email FROM users WHERE id IN {SqliteSql.InList([.. _hundredItems])}").ToList();
    }
}

File: fsharp/benchmarks/StringThing.FSharp.Benchmarks/InListBenchmarks.fs
namespace StringThing.FSharp.Benchmarks

open BenchmarkDotNet.Attributes
open Microsoft.Data.Sqlite
open StringThing.FSharp
open StringThing.FSharp.Sqlite

type InListUser =
    { Id: int64
      Name: string
      Email: string option }

module InListUser =
    let row : RowReader<InListUser> =
        row {
            let! id    = Row.int64 "id"
            and! name  = Row.string "name"
            and! email = Row.stringOption "email"
            return { Id = id; Name = name; Email = email }
        }

[<MemoryDiagnoser>]
type InListBenchmarks() =

    let mutable connection : SqliteConnection = Unchecked.defaultof<_>
    let mutable tenItems : int64[] = Unchecked.defaultof<_>
    let mutable hundredItems : int64[] = Unchecked.defaultof<_>

    let prebuiltSql10 =
        "SELECT id, name, email FROM users WHERE id IN ("
        + System.String.Join(", ", [| for i in 0..9 -> sprintf "@p%d" i |])
        + ")"

    let prebuiltSql100 =
        "SELECT id, name, email FROM users WHERE id IN ("
        + System.String.Join(", ", [| for i in 0..99 -> sprintf "@p%d" i |])
        + ")"

    [<GlobalSetup>]
    member _.Setup() =
        connection <- new SqliteConnection("Data Source=:memory:")
        connection.Open()

        use create = connection.CreateCommand()
        create.CommandText <-
            "CREATE TABLE users ( \
                id INTEGER PRIMARY KEY, \
                name TEXT NOT NULL, \
                email TEXT)"
        create.ExecuteNonQuery() |> ignore

        for i = 1 to 100 do
            use insert = connection.CreateCommand()
            insert.CommandText <-
                sprintf "INSERT INTO users (id, name, email) VALUES (%d, 'user_%d', 'user_%d@example.com')" i i i
            insert.ExecuteNonQuery() |> ignore

        tenItems <- [| 1L .. 10L |]
        hundredItems <- [| 1L .. 100L |]

    [<GlobalCleanup>]
    member _.Cleanup() =
        connection.Dispose()

    [<Benchmark(Description = "Raw: IN 10 (prebuilt SQL)")>]
    member _.Raw_InList_10() : System.Collections.Generic.List<InListUser> =
        use cmd = connection.CreateCommand()
        cmd.CommandText <- prebuiltSql10
        for i = 0 to 9 do
            cmd.Parameters.AddWithValue(sprintf "@p%d" i, tenItems.[i]) |> ignore
        use reader = cmd.ExecuteReader()
        let results = System.Collections.Generic.List<InListUser>()
        while reader.Read() do
            results.Add({
                Id = reader.GetInt64(0)
                Name = reader.GetString(1)
                Email = if reader.IsDBNull(2) then None else Some (reader.GetString(2))
            })
        results

    [<Benchmark(Description = "FSharp: IN 10")>]
    member _.FSharp_InList_10() : InListUser list =
        let ids = Sqlite.inList tenItems
        connection.QueryString InListUser.row $"""
            SELECT id, name, email
            FROM users
            WHERE id IN {ids}
            """
        |> Seq.toList

    [<Benchmark(Description = "Raw: IN 100 (prebuilt SQL)")>]
    member _.Raw_InList_100() : System.Collections.Generic.List<InListUser> =
        use cmd = connection.CreateCommand()
        cmd.CommandText <- prebuiltSql100
        for i = 0 to 99 do
            cmd.Parameters.AddWithValue(sprintf "@p%d" i, hundredItems.[i]) |> ignore
        use reader = cmd.ExecuteReader()
        let results = System.Collections.Generic.List<InListUser>()
        while reader.Read() do
            results.Add({
                Id = reader.GetInt64(0)
                Name = reader.GetString(1)
                Email = if reader.IsDBNull(2) then None else Some (reader.GetString(2))
            })
        results

    [<Benchmark(Description = "FSharp: IN 100")>]
    member _.FSharp_InList_100() : InListUser list =
        let ids = Sqlite.inList hundredItems
        connection.QueryString InListUser.row $"""
            SELECT id, name, email
            FROM users
            WHERE id IN {ids}
            """
        |> Seq.toList
